src/lib/__tests__/InstantSearch-test.js

```diff
@@ -151,742 +162,890 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/instantsear
+
+    expect(widget.render).toHaveBeenCalledTimes(1);
+
+    // Trigger multiple searches
+    search.mainHelper.search();
+    search.mainHelper.search();
+    search.mainHelper.search();
+    search.mainHelper.search();
+
+    // Reaches the delay
+    jest.runOnlyPendingTimers();
 
-    done();
+    // Wait for the `render`
+    await runAllMicroTasks();
+
+    expect(widget.render).toHaveBeenCalledTimes(2);
   });
 
-  it('should set the DerivedHelper to `null`', () => {
+  it('triggers a `render` once the search expires the delay', async () => {
+    const { searches, searchClient } = createControlledSearchClient();
     const search = new InstantSearch({
-      indexName: '',
-      searchClient: {
-        search() {
-          return Promise.resolve({
-            results: [
-              {
-                query: 'fake query',
-              },
-            ],
-          });
-        },
-      },
+      indexName: 'index_name',
+      searchClient,
     });
 
+    const widget = createWidget();
+
+    search.addWidget(widget);
+
+    expect(widget.render).toHaveBeenCalledTimes(0);
+
     search.start();
 
-    expect(search.derivedHelper).not.toBe(null);
+    expect(widget.render).toHaveBeenCalledTimes(0);
 
-    search.dispose();
+    // Resolve the `search`
+    searches[0].resolver();
 
-    expect(search.derivedHelper).toBe(null);
+    // Wait for the `search`
+    await searches[0].promise;
 
-    search.start();
+    // Wait for the `render`
+    await runAllMicroTasks();
+
+    expect(widget.render).toHaveBeenCalledTimes(1);
+    expect(widget.render).toHaveBeenLastCalledWith(
+      expect.objectContaining({
+        searchMetadata: {
+          isSearchStalled: false,
+        },
+      })
+    );
+
+    // Trigger a new search
+    search.mainHelper.search();
+
+    expect(widget.render).toHaveBeenCalledTimes(1);
+
+    // The delay is reached
+    jest.runOnlyPendingTimers();
+
+    // Wait for the `render`
+    await runAllMicroTasks();
+
+    // Widgets render because of the stalled search
+    expect(widget.render).toHaveBeenCalledTimes(2);
+    expect(widget.render).toHaveBeenLastCalledWith(
+      expect.objectContaining({
+        searchMetadata: {
+          isSearchStalled: true,
+        },
+      })
+    );
 
-    expect(search.derivedHelper).not.toBe(null);
+    // Resolve the `search`
+    searches[1].resolver();
+
+    // Wait for the `search`
+    await searches[1].promise;
+
+    // Wait for the `render`
+    await runAllMicroTasks();
+
+    // Widgets render because of the results
+    expect(widget.render).toHaveBeenCalledTimes(3);
+    expect(widget.render).toHaveBeenLastCalledWith(
+      expect.objectContaining({
+        searchMetadata: {
+          isSearchStalled: false,
+        },
+      })
+    );
   });
 
-  it('should detach the DerivedHelper', () => {
+  // https://github.com/algolia/instantsearch.js/pull/2623
+  it('does not trigger a re-`render` without results', () => {
```

Comment:
This feature is meant for InstantSearch to not render anything until it receives some results, right? (only the first load the first time)

Reply:
Yes, otherwise on the first `search` InstantSearch schedules a stalled render. If the delay is reached we call the render, but we don't have the results yet which throws an error.